mesh/mesh.go

```diff
@@ -172,31 +152,30 @@ func (msh *Mesh) LatestLayer() types.LayerID {
 //
 // First valid layer starts with 1. 0 is empty layer and can be ignored.
 func (msh *Mesh) MissingLayer() types.LayerID {
-	msh.mutex.RLock()
-	defer msh.mutex.RUnlock()
-	return msh.missingLayer
-}
-
-func (msh *Mesh) setMissingLayer(lid types.LayerID) {
-	msh.mutex.Lock()
-	defer msh.mutex.Unlock()
-	msh.missingLayer = lid
+	value := msh.missingLayer.Load()
+	if value == nil {
+		return types.LayerID{}
+	}
+	return value.(types.LayerID)
 }
 
 // setLatestLayer sets the latest layer we saw from the network.
-func (msh *Mesh) setLatestLayer(idx types.LayerID) {
+func (msh *Mesh) setLatestLayer(lid types.LayerID) {
 	events.ReportLayerUpdate(events.LayerUpdate{
-		LayerID: idx,
+		LayerID: lid,
 		Status:  events.LayerStatusTypeUnknown,
 	})
-	defer msh.mutex.Unlock()
-	msh.mutex.Lock()
-	if idx.After(msh.latestLayer) {
-		events.ReportNodeStatusUpdate()
-		msh.With().Info("set latest known layer", idx)
-		msh.latestLayer = idx
-		if err := layers.SetStatus(msh.db, idx, layers.Latest); err != nil {
-			msh.Error("could not persist latest layer index")
+	for {
```

Comment:
what is the purpose of the for block?

Reply:
yikes. i don't think i understand this. how does the code exit the for loop?

Reply:
if there are two concurrent updates with layer 9 and layer 10, it is important not execute set=10 and then set=9.  in order to guarantee it i am using CAS on the previous value. for example two threads:

- T1. lid = 9
- T2. lid = 10
- T1. load. current = 8
- T2. load. current = 8
- T2. update. CAS 8 10. SUCCESS => exit
- T1. update CAS 8 9. FAILED
- T1. load. current = 10
- T1.  lid <= current => exit
-

Reply:
i see. thanks for the explanation. 
tho is this extra logic still necessary given that you always hold a write lock when executing this?

Reply:
setLatestLayer is called when block or proposal are added